### DISCUSSION TITLE
ORM Mapper cannot map a table which does not have primary key but has UniqueConstraint

### DISCUSSION THREAD

**Discussion by fregataa** [sqlalchemy/sqlalchemy#8153] [Usage Questions]
```
import sqlalchemy as sa

group_users = sa.Table(
    'group_users', metadata,
    sa.Column('user_id', GUID,
              sa.ForeignKey('users.uuid', onupdate='CASCADE', ondelete='CASCADE'),
              nullable=False),
    sa.Column('group_id', GUID,
              sa.ForeignKey('groups.id', onupdate='CASCADE', ondelete='CASCADE'),
              nullable=False),
    sa.UniqueConstraint('user_id', 'group_id', name='uq_association_user_id_group_id'),
)
```
I have an imperative mapped table like above.
and I try to make a hybrid table like below.

```
class GroupUsers(Base):
    __table__ = group_users
    __table_args__ = (
        sa.PrimaryKeyConstraint('user_id', 'group_id', name='pk_association_user_id_group_id'),
    )
```

I add a PrimaryKeyConstraint in `__table_args__` since ORM need any kind of primary key stuff.
but this table occurs error like this.

```
...
File "/home/python/virtualenvs/python-default/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 665, in __init__
    self._configure_pks()
  File "/home/python/virtualenvs/python-default/3.10.4/lib/python3.10/site-packages/sqlalchemy/orm/mapper.py", line 1360, in _configure_pks
    raise sa_exc.ArgumentError(
sqlalchemy.exc.ArgumentError: Mapper mapped class GroupUsers->group_users could not assemble any primary key columns for mapped table 'group_users'
```

I think that `_.orm.Mapper` class should gracefully configure pks cause it seems that `_orm.Mapper` class does not consider `__table_args__`
or is there any way to override constraints of inherited table -in this case, `group_users`- ?

**Comment by fregataa** [comment]
I finally understand.
it is impossible, since the `group_users` table has only unique constraint on DB schema, not primary constraint.
I need to make a migrate `UniqueConstraint` to `PrimaryKeyConstraint` before create new declarative map table!

**Comment by zzzeek** [comment]
hi -

make use of the [primary_key](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.mapper.params.primary_key) parameter within the [__mapper_args__](https://docs.sqlalchemy.org/en/14/orm/declarative_config.html#mapper-configuration-options-with-declarative) dictionary.   not too much examples for this but there is one at [Mapping a subset of table columns](https://docs.sqlalchemy.org/en/14/orm/mapping_columns.html#mapping-a-subset-of-table-columns)


am working on cleanup of ORM docs right now
